sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/ReleaseDownloadListener.java

```diff
@@ -0,0 +1,196 @@
+
+/**
+ * Listener for downloading progress.
+ */
+class ReleaseDownloadListener implements ReleaseDownloader.Listener {
+
+    /**
+     * Context.
+     */
+    private final Context mContext;
+
+    /**
+     * Private field to store information about release we are currently working with.
+     */
+    private final ReleaseDetails mReleaseDetails;
+
+    ReleaseDownloadListener(@NonNull Context context, @NonNull ReleaseDetails releaseDetails) {
+        mContext = context;
+        mReleaseDetails = releaseDetails;
+    }
+
+    /**
+     * Last download progress dialog that was shown.
+     * Android 8 deprecates this dialog but only reason is that they want us to use a non modal
+     * progress indicator while we actually use it to be a modal dialog for forced update.
+     * They will always keep this dialog to remain compatible but just mark it deprecated.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    private android.app.ProgressDialog mProgressDialog;
+
+    @Override
+    public void onStart(long enqueueTime) {
+        AppCenterLog.debug(LOG_TAG, String.format(Locale.ENGLISH, "Start download %s (%d) update.",
+                mReleaseDetails.getShortVersion(), mReleaseDetails.getVersion()));
+        SharedPreferencesManager.putInt(PREFERENCE_KEY_DOWNLOAD_STATE, DOWNLOAD_STATE_ENQUEUED);
+        SharedPreferencesManager.putLong(PREFERENCE_KEY_DOWNLOAD_TIME, enqueueTime);
+    }
+
+    @Override
+    @WorkerThread
+    public boolean onProgress(final long currentSize, final long totalSize) {
+        AppCenterLog.verbose(LOG_TAG, String.format(Locale.ENGLISH, "Downloading %s (%d) update: %d KiB / %d KiB",
+                mReleaseDetails.getShortVersion(), mReleaseDetails.getVersion(),
+                currentSize / KIBIBYTE_IN_BYTES, totalSize / KIBIBYTE_IN_BYTES));
+
+        /* If file size is known update downloadProgress bar. */
+        if (mProgressDialog != null && totalSize >= 0) {
+            HandlerUtils.runOnUiThread(new Runnable() {
+
+                @Override
+                @SuppressWarnings({"deprecation", "RedundantSuppression"})
+                public void run() {
+
+                    /* When we switch from indeterminate to determinate */
+                    if (mProgressDialog.isIndeterminate()) {
+
+                        /* Configure the progress dialog determinate style. */
+                        mProgressDialog.setProgressPercentFormat(NumberFormat.getPercentInstance());
+                        mProgressDialog.setProgressNumberFormat(mContext.getString(R.string.appcenter_distribute_download_progress_number_format));
+                        mProgressDialog.setIndeterminate(false);
+                        mProgressDialog.setMax((int) (totalSize / MEBIBYTE_IN_BYTES));
+                    }
+                    mProgressDialog.setProgress((int) (currentSize / MEBIBYTE_IN_BYTES));
+                }
+            });
+            return true;
```

Comment:
The file size might not be known right away with download manager (it needs to actually start downloading for that), yet we still need to check later. But here we return false if totalSize <= 0. How will we update from indeterminate to determinate in that case if we are not checking again later? (for android 5+).

Reply:
I checked develop we were starting the timer always (as in total size 0 or not).

Reply:
Remove this return, and for the other return instead of false, return mProgressDialog != null (or release is mandatory check).